package com.joseetoon.emfplayeranimatorbridge.mixin.client;

import com.joseetoon.emfplayeranimatorbridge.compat.Compatibility;
import com.joseetoon.emfplayeranimatorbridge.compat.EmfAnimationBridge;
import com.mojang.blaze3d.vertex.PoseStack;
import net.minecraft.client.model.EntityModel;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.entity.EntityRenderer;
import net.minecraft.client.renderer.entity.EntityRendererProvider;
import net.minecraft.client.renderer.entity.LivingEntityRenderer;
import net.minecraft.client.renderer.entity.RenderLayerParent;
import net.minecraft.world.entity.LivingEntity;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@Mixin(value = LivingEntityRenderer.class, priority = 1100)
public abstract class LivingEntityRendererResumeMixin<
        T extends LivingEntity,
        M extends EntityModel<T>>
        extends EntityRenderer<T>
        implements RenderLayerParent<T, M> {
    protected LivingEntityRendererResumeMixin(EntityRendererProvider.Context context) {
        super(context);
    }

    @Inject(
            method = "render(Lnet/minecraft/world/entity/LivingEntity;FFLcom/mojang/blaze3d/vertex/PoseStack;Lnet/minecraft/client/renderer/MultiBufferSource;I)V",
            at = @At("TAIL")
    )
    protected void emfplayeranimatorbridge$postRenderToBuffer(T entity, float entityYaw, float tickDelta, PoseStack poseStack,
                                                             MultiBufferSource bufferSource, int packedLight, CallbackInfo ci) {
        if (Compatibility.isEmfAvailable()) {
            EmfAnimationBridge.postRenderToBuffer(entity, this.getModel());
        }
    }
}
